src/includes/dinit-iostream.h

```diff
@@ -0,0 +1,742 @@
+ * 4. A "flush" object is passed to write() functions (equivalent to calling flush()).
+ * 5. An "endl" object is passed to write() functions (equivalent to writing \n and
+ *    calling flush()).
+ * 6. The object is destructed (which calls flush_nothrow()).
+ *
+ * write() functions guarantee to return the number of actually written characters,
+ * not the number of buffered characters. They can also return -1 on failure cases.
+ *
+ * Public member: flush()
+ * ----------------------
+ *
+ * The ostream provides a flush() function to write all contents of the buffer.
+ * Passing a "flush" object to write() functions is equivalent to calling flush()
+ * on the stream. This function will return the number of all written characters
+ * or -1 on failure cases.
+ *
+ * Error handling and Stream states
+ * --------------------------------
+ *
+ * ostream states are: buffer_failbit and posixio_errno.
+ * buffer_failbit is false by default and means the class constructor failed to
+ * allocate a unique_ptr of the buffer at construction time.
+ * posixio_errno is 0 by default and means there is something wrong with POSIX I/O calls,
+ * It's set to the POSIX errno in this case.
+ *
+ * For checking current stream states, these functions are available:
+ * 1. rdstate(): Returns all current state bits or 0 when everything looks good.
+ * 2. good(): Returns 1 when there are no set bits otherwise 0.
+ * 3. buffer_failure(): Returns 1 or 0 based on whether buffer_failbit is set or not.
+ * 4. posixio_failure(): Returns 0 or the POSIX errno based on the posixio_errno value.
+ *
+ * For setting current stream state(s), these functions are available:
+ * 1. clear(): Sets all bits/vars to zero or false.
+ * 2. setstate(): Sets requested bits/vars to the requested value.
+ *
+ * Exception handling
+ * ------------------
+ *
+ * throw_on_states() can be used to set when an exception should be thrown or not.
+ * For ostream, it accepts diostates::buffer_failbit and diostates::posixio_errno.
+ *
+ * clear_throws() can be used for clear all marked situations for throwing exceptions.
+ *
+ * See the description for each function to know what exceptions can be thrown by
+ * functions.
+ *
+ * All exception-throwing functions have exception-free variants marked by the
+ * "_nothrow" suffix. Those functions guarantee not to throw any exceptions
+ * regardless of the throw_on_states() setting.
+ */
+class ostream : public streambuf_base
+{
+    int fd = -1;
+
+    // Variables to capture stream current status:
+    bool buffer_failbit = false;
+    // posixio_errno has different functionality, it will be set to POSIX errno
+    // which returned by POSIX functions (such as read()).
+    int posixio_errno = 0;
+
+    // Keep when should throw an exception (on which case):
+    bool throw_on_buffer_fail = false;
+    bool throw_on_posixio_fail = false;
+
+    // Internal function to appending into the buffer
+    ssize_t put(const char *msg) noexcept;
```

Comment:
Hmm, Do you mean there is should be a `write_binary()` or `write_raw()` function?

Reply:
No, I really meant that `put` and the various overloads of `write` should work with binary data as well as nul-terminated strings. If you use just a `char *` to identify the data then you can only handle nul-terminated strings.

The current `write(const std::string &)` overload accepts a string, which may contain nul characters. Let's say you try to use it to write a string which is `"hello\0hello"` (11 chars in length, '\0' is  a nul). It gets converted to a C-style string (`char *`) and passed to `put()`, which calls `strlen()` - which will return 5 characters, not 11, because the nul character terminates a C string. So in that case only 5 characters will be written, but all 11 should have been written.

So to fix that, add another argument to `put` which is the number of chars to write, instead of using `strlen` inside the `put` function.